### PR TITLE
Apply clang-tidy recommendations to component 'cdi_analytic'

### PR DESCRIPTION
**Pull request by KineticTheory**
### Description of changes

+ Convert `typedef` commands to `using`.
+ Decorate member functions of derived classes as `override`.
+ Make use of `= default` for destructors.
+ Use `std::make_shared` instead of `.reset`.
+ Use `auto` for variable declarations when the type is specified as part of the assignment.
+ Use initializer lists.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [ ] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation
* WIP
  * [x] Requires a rebase after #936 is merged.

### CODE REVIEW

**Comment by KineticTheory** [comment]
@attom Can you review these changes and merge the PR?  The changes are clang-tidy transformations.  The coverage numbers did drop because I removed about 180 lines of code - but no tests were actually removed.

File: src/cdi_analytic/Analytic_EICoupling.cc [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace rtt_cdi_analytic {
  *        rtt_cdi_analytic::Analytic_EICoupling_Model object
  */
 Analytic_EICoupling::Analytic_EICoupling(SP_Analytic_Model model_in)
-    : analytic_model(model_in) {
+    : analytic_model(std::move(model_in)) {
```
Comment (attom) [inline src/cdi_analytic/Analytic_EICoupling.cc:29]: the move is safe in this case?

File: src/cdi_analytic/Analytic_Gray_Opacity.cc [hunk 2]
```diff
@@ -33,7 +33,7 @@ namespace rtt_cdi_analytic {
 Analytic_Gray_Opacity::Analytic_Gray_Opacity(SP_Analytic_Model model_in,
                                              rtt_cdi::Reaction reaction_in,
                                              rtt_cdi::Model cdi_model_in)
-    : analytic_model(model_in), reaction(reaction_in), model(cdi_model_in) {
+    : analytic_model(std::move(model_in)), reaction(reaction_in), model(cdi_model_in) {
```
Comment (attom) [inline src/cdi_analytic/Analytic_Gray_Opacity.cc:36]: I guess it's not clear why model_in can be moved but reaction_in and cdi_model_in are copied

File: src/cdi_analytic/Analytic_MultigroupOpacity.cc [hunk 3]
```diff
@@ -29,26 +29,26 @@ namespace rtt_cdi_analytic {
...
  * \param[in] model_in CDI type
  */
-Analytic_MultigroupOpacity::Analytic_MultigroupOpacity(
-    const sf_double &groups, rtt_cdi::Reaction reaction_in,
-    rtt_cdi::Model model_in)
-    : group_boundaries(groups), reaction(reaction_in), model(model_in) {
+Analytic_MultigroupOpacity::Analytic_MultigroupOpacity(sf_double groups,
+                                                       rtt_cdi::Reaction reaction_in,
+                                                       rtt_cdi::Model model_in)
+    : group_boundaries(std::move(groups)), reaction(reaction_in), model(model_in) {
```
Comment (attom) [inline src/cdi_analytic/Analytic_MultigroupOpacity.cc:40]: another move question

File: src/cdi_analytic/Analytic_EICoupling.cc (hunk 1 above)
Comment (attom) [reply src/cdi_analytic/Analytic_EICoupling.cc:29]: is it safe because you're copying in anyway, so you might as well move in the copy?

Comment (alexrlongne) [reply src/cdi_analytic/Analytic_EICoupling.cc:29]: That's right! Sorry to chime in, @keadyk and I were looking at this yesterday in another clang-tidy change. It's an optimization in certain cases to allow the compiler to use the move constructor. Explained better here:
https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

Comment (attom) [reply src/cdi_analytic/Analytic_EICoupling.cc:29]: Thanks! As I thought more about, it made more sense, but definitely a new pattern for me to learn. The article was super helpful, gracias!

File: src/cdi_analytic/Analytic_MultigroupOpacity.cc (hunk 3 above)
Comment (attom) [reply src/cdi_analytic/Analytic_MultigroupOpacity.cc:40]: I don't like that `std::vector<double>` is the somewhat cryptic "sf_double", but I'll live

**Comment by attom** [comment]
@KineticTheory, would you like me to merge? My only dislike is the typename 'sf_double', which I find to be non-intuitive.

**Comment by KineticTheory** [comment]
> @KineticTheory, would you like me to merge? My only dislike is the typename 'sf_double', which I find to be non-intuitive.

Please merge this.

I'm with you on the `sf_double` syntax; others disagree.  BTW - it stands for _scalar-field-double_ as opposed to a `vf_double` meaning _vector-field-double_.

**Comment by KineticTheory** [comment]
@attom Rebase was required. Let's let the CI run again.

**Comment by attom** [comment]
Will merge once CI runs again

**Comment by attom** [comment]
Something seems amiss. @KineticTheory, should I just go ahead and merge or would you like to restart the CI? (Or should I just be more patient?)

**Comment by KineticTheory** [comment]
> Something seems amiss. @KineticTheory, should I just go ahead and merge or would you like to restart the CI? (Or should I just be more patient?)

Please merge it.  Travis looks broken. If you click on Appveyor details, you will see that it is done and tests pass.

**Pull request merged**